File: tools/curve-qc.py
#!/usr/bin/env python3
"""
Read LAS with lasio, compute RMSE and NRMSE for a given curve.
Usage: python curve-qc.py file.las CURVE=GR
"""

import sys
import os
import json
from typing import Dict, List, Optional

def compute_rmse_nrmse(values: List[float], fitted_values: List[float]) -> Dict[str, float]:
    """
    Compute Root Mean Square Error (RMSE) and Normalized RMSE (NRMSE)
    """
    if len(values) != len(fitted_values) or len(values) == 0:
        return {"rmse": float('nan'), "nrmse": float('nan')}
    
    # Remove NaN values
    valid_pairs = [(v, f) for v, f in zip(values, fitted_values) 
                   if not (v != v or f != f)]  # NaN check
    
    if len(valid_pairs) == 0:
        return {"rmse": float('nan'), "nrmse": float('nan')}
    
    values_clean, fitted_clean = zip(*valid_pairs)
    
    # Calculate RMSE
    squared_errors = [(v - f) ** 2 for v, f in valid_pairs]
    mse = sum(squared_errors) / len(squared_errors)
    rmse = mse ** 0.5
    
    # Calculate NRMSE (normalized by range)
    value_range = max(values_clean) - min(values_clean)
    nrmse = rmse / value_range if value_range > 0 else float('nan')
    
    return {"rmse": rmse, "nrmse": nrmse}

def analyze_las_curve(file_path: str, curve_name: str) -> Dict:
    """
    Analyze a specific curve in a LAS file
    """
    try:
        # Try to import lasio, fall back to basic parsing if not available
        try:
            import lasio
            las = lasio.read(file_path)
            
            # Get curve data
            if curve_name not in las.curves:
                available_curves = [c.mnemonic for c in las.curves]
                return {
                    "error": f"Curve '{curve_name}' not found. Available curves: {available_curves}",
                    "available_curves": available_curves
                }
            
            curve_data = las[curve_name]
            depth_data = las.index
            
            # Basic statistics
            valid_data = [v for v in curve_data if v == v]  # Remove NaN
            
            if len(valid_data) == 0:
                return {"error": f"No valid data points for curve '{curve_name}'"}
            
            stats = {
                "curve": curve_name,
                "total_points": len(curve_data),
                "valid_points": len(valid_data),
                "min_value": min(valid_data),
                "max_value": max(valid_data),
                "mean_value": sum(valid_data) / len(valid_data),
                "depth_start": float(depth_data.min()),
                "depth_stop": float(depth_data.max()),
                "depth_step": float(las.well.STEP.value) if las.well.STEP else None
            }
            
            # Create simple fitted curve (linear trend for demo)
            fitted_values = []
            if len(valid_data) > 1:
                # Simple linear fit for demonstration
                slope = (valid_data[-1] - valid_data[0]) / len(valid_data)
                for i in range(len(curve_data)):
                    fitted_values.append(valid_data[0] + slope * i)
                
                # Compute QC metrics
                qc_metrics = compute_rmse_nrmse(list(curve_data), fitted_values)
                stats.update(qc_metrics)
            
            return stats
            
        except ImportError:
            # Fallback: basic LAS parsing without lasio
            return parse_las_basic(file_path, curve_name)
            
    except Exception as e:
        return {"error": f"Failed to analyze LAS file: {str(e)}"}

def parse_las_basic(file_path: str, curve_name: str) -> Dict:
    """
    Basic LAS parsing fallback when lasio is not available
    """
    with open(file_path, 'r') as f:
        content = f.read()
    
    lines = content.split('\n')
    curves = []
    data_section = False
    
    # Find curve definitions
    for line in lines:
        line = line.strip()
        if line.startswith('~C'):
            data_section = False
            continue
        elif line.startsWith('~A'):
            data_section = True
            break
        elif not data_section and '.' in line and not line.startswith('#'):
            curve_mnemonic = line.split('.')[0].strip()
            if curve_mnemonic:
                curves.append(curve_mnemonic)
    
    if curve_name not in curves:
        return {
            "error": f"Curve '{curve_name}' not found. Available curves: {curves}",
            "available_curves": curves
        }
    
    # For basic demo, return minimal stats
    return {
        "curve": curve_name,
        "message": "Basic parsing - install lasio for full functionality",
        "available_curves": curves,
        "rmse": 0.0,
        "nrmse": 0.0
    }

def main():
    if len(sys.argv) < 3:
        print("Usage: python curve-qc.py <file.las> CURVE=<curve_name>", file=sys.stderr)
        print("Example: python curve-qc.py demo.las CURVE=GR", file=sys.stderr)
        sys.exit(1)
    
    file_path = sys.argv[1]
    curve_arg = sys.argv[2]
    
    if not os.path.exists(file_path):
        print(f"Error: File '{file_path}' not found", file=sys.stderr)
        sys.exit(1)
    
    if not curve_arg.startswith('CURVE='):
        print("Error: Second argument must be CURVE=<curve_name>", file=sys.stderr)
        sys.exit(1)
    
    curve_name = curve_arg.split('=')[1]
    
    # Analyze the curve
    result = analyze_las_curve(file_path, curve_name)
    
    # Output JSON result
    print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()

File: tools/curve-qc.ts
#!/usr/bin/env tsx
/**
 * SHALE YEAH Curve Quality Control (TypeScript)
 * 
 * Computes RMSE and NRMSE for curve analysis and validation
 * Replaces curve-qc.py with modern TypeScript implementation
 */

import fs from "node:fs";
import { parseLASFile, type LASData } from './las-parse.js';

interface QCMetrics {
  rmse: number;
  nrmse: number;
}

interface CurveAnalysis {
  curve: string;
  totalPoints: number;
  validPoints: number;
  minValue: number;
  maxValue: number;
  meanValue: number;
  depthStart: number;
  depthStop: number;
  depthStep: number;
  rmse?: number;
  nrmse?: number;
  error?: string;
  availableCurves?: string[];
}

/**
 * Compute Root Mean Square Error (RMSE) and Normalized RMSE (NRMSE)
 */
function computeRMSE_NRMSE(values: number[], fittedValues: number[]): QCMetrics {
  if (values.length !== fittedValues.length || values.length === 0) {
    return { rmse: NaN, nrmse: NaN };
  }

  // Remove NaN values
  const validPairs: [number, number][] = [];
  for (let i = 0; i < values.length; i++) {
    const v = values[i];
    const f = fittedValues[i];
    if (!isNaN(v) && !isNaN(f)) {
      validPairs.push([v, f]);
    }
  }

  if (validPairs.length === 0) {
    return { rmse: NaN, nrmse: NaN };
  }

  // Calculate RMSE
  let sumSquaredErrors = 0;
  const validValues: number[] = [];
  
  for (const [v, f] of validPairs) {
    sumSquaredErrors += (v - f) ** 2;
    validValues.push(v);
  }

  const mse = sumSquaredErrors / validPairs.length;
  const rmse = Math.sqrt(mse);

  // Calculate NRMSE (normalized by range)
  const minValue = Math.min(...validValues);
  const maxValue = Math.max(...validValues);
  const valueRange = maxValue - minValue;
  const nrmse = valueRange > 0 ? rmse / valueRange : NaN;

  return { rmse, nrmse };
}

/**
 * Create a simple fitted curve using linear trend (for demonstration)
 */
function createLinearFit(values: number[]): number[] {
  if (values.length <= 1) {
    return values.slice();
  }

  // Filter out NaN values to calculate slope
  const validValues = values.filter(v => !isNaN(v));
  if (validValues.length < 2) {
    return values.slice();
  }

  const slope = (validValues[validValues.length - 1] - validValues[0]) / validValues.length;
  const startValue = validValues[0];

  return values.map((_, i) => startValue + slope * i);
}

/**
 * Analyze a specific curve in a LAS file
 */
function analyzeLASCurve(filePath: string, curveName: string): CurveAnalysis {
  try {
    // Parse LAS file using our existing parser
    const lasData: LASData = parseLASFile(filePath);
    
    // Find the requested curve
    const curve = lasData.curves.find(c => c.name.toUpperCase() === curveName.toUpperCase());
    
    if (!curve) {
      const availableCurves = lasData.curves.map(c => c.name);
      return {
        curve: curveName,
        error: `Curve '${curveName}' not found. Available curves: ${availableCurves.join(', ')}`,
        availableCurves,
        totalPoints: 0,
        validPoints: 0,
        minValue: 0,
        maxValue: 0,
        meanValue: 0,
        depthStart: 0,
        depthStop: 0,
        depthStep: 0
      };
    }

    // Get valid data points (filter out NaN)
    const validData = curve.data.filter(v => !isNaN(v));
    
    if (validData.length === 0) {
      return {
        curve: curveName,
        error: `No valid data points for curve '${curveName}'`,
        totalPoints: curve.data.length,
        validPoints: 0,
        minValue: 0,
        maxValue: 0,
        meanValue: 0,
        depthStart: lasData.depth_start,
        depthStop: lasData.depth_stop,
        depthStep: lasData.depth_step
      };
    }

    // Calculate basic statistics
    const minValue = Math.min(...validData);
    const maxValue = Math.max(...validData);
    const meanValue = validData.reduce((sum, v) => sum + v, 0) / validData.length;

    const analysis: CurveAnalysis = {
      curve: curveName,
      totalPoints: curve.data.length,
      validPoints: validData.length,
      minValue,
      maxValue,
      meanValue,
      depthStart: lasData.depth_start,
      depthStop: lasData.depth_stop,
      depthStep: lasData.depth_step
    };

    // Create fitted curve and compute QC metrics
    if (validData.length > 1) {
      const fittedValues = createLinearFit(curve.data);
      const qcMetrics = computeRMSE_NRMSE(curve.data, fittedValues);
      
      analysis.rmse = qcMetrics.rmse;
      analysis.nrmse = qcMetrics.nrmse;
    } else {
      analysis.rmse = 0;
      analysis.nrmse = 0;
    }

    return analysis;

  } catch (error) {
    return {
      curve: curveName,
      error: `Failed to analyze LAS file: ${error}`,
      totalPoints: 0,
      validPoints: 0,
      minValue: 0,
      maxValue: 0,
      meanValue: 0,
      depthStart: 0,
      depthStop: 0,
      depthStep: 0
    };
  }
}

/**
 * CLI usage
 */
const main = () => {
  const args = process.argv.slice(2);
  
  if (args.length < 2) {
    console.error('Usage: curve-qc <file.las> CURVE=<curve_name>');
    console.error('Example: curve-qc demo.las CURVE=GR');
    process.exit(1);
  }

  const filePath = args[0];
  const curveArg = args[1];

  if (!fs.existsSync(filePath)) {
    console.error(`Error: File '${filePath}' not found`);
    process.exit(1);
  }

  if (!curveArg.startsWith('CURVE=')) {
    console.error('Error: Second argument must be CURVE=<curve_name>');
    process.exit(1);
  }

  const curveName = curveArg.split('=')[1];
  
  if (!curveName) {
    console.error('Error: Curve name cannot be empty');
    process.exit(1);
  }

  try {
    const result = analyzeLASCurve(filePath, curveName);
    console.log(JSON.stringify(result, null, 2));
  } catch (error) {
    console.error(`Error analyzing curve: ${error}`);
    process.exit(1);
  }
};

// ESM compatible check
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { analyzeLASCurve, computeRMSE_NRMSE, type CurveAnalysis, type QCMetrics };